File: frontend/src/lib/models/variable_config.ts
export interface VariableConfig {
    id: string;
    type: string;
}

export let VARIABLE_TYPE_DIRECT = "direct";

export interface DirectVariableConfig extends VariableConfig {
    value: string;
}

export let VARIABLE_TYPE_REQUEST = "request";

export interface RequestVariableConfig extends VariableConfig {
    lazy?: boolean;
    url: string;
    method?: string;
    body?: string;
    headers?: { [key: string]: string };
    placeholder?: string;
}

File: src/main/kotlin/dev/schlaubi/dashboard/models/VariableConfig.kt
package dev.schlaubi.dashboard.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Variable definition
@Serializable
sealed class VariableConfig {
    abstract val id: String
}

@Serializable
@SerialName("direct")
data class DirectVariableConfig(
    override val id: String, val value: String
) : VariableConfig()

@Serializable
@SerialName("request")
data class RequestVariableConfig(
    override val id: String,
    val lazy: Boolean? = null,
    val url: String,
    val method: String? = null,
    val body: String? = null,
    val headers: Map<String, String>? = null,
    val placeholder: String? = null
) : VariableConfig()